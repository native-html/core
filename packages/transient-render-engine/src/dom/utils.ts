import { Text as TextType, Tag as TagType } from 'domelementtype';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === TextType;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === TagType;
}
