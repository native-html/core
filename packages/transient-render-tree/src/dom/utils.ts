import { ElementType } from 'htmlparser2';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === ElementType.Text;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === ElementType.Tag;
}
