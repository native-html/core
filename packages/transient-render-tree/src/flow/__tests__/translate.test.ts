import { rfc002Source, href, imgSrc } from './shared';
import { translateTreeTest } from './utils';

describe('translateNode function', () => {
  it('should comply with RFC002 example (translating)', async () => {
    const ttree = await translateTreeTest(rfc002Source);
    expect(ttree).toMatchObject({
      type: 'phrasing',
      isAnchor: true,
      attributes: { href },
      children: [
        {
          type: 'text',
          attributes: {},
          tagName: null,
          data: '\nThis is\n'
        },
        {
          type: 'text',
          attributes: {},
          tagName: 'span',
          data: 'phrasing content'
        },
        {
          type: 'text',
          attributes: {},
          tagName: null,
          data: '\n'
        },
        {
          type: 'block',
          attributes: { src: imgSrc },
          tagName: 'img'
        },
        {
          type: 'text',
          attributes: {},
          tagName: null,
          data: '\n    and this is '
        },
        {
          type: 'text',
          attributes: {},
          tagName: 'strong',
          data: 'too'
        },
        {
          type: 'text',
          attributes: {},
          tagName: null,
          data: '.\n'
        }
      ]
    });
    expect(ttree).toMatchSnapshot();
  });
  it('should translate styles', async () => {
    const ttree = await translateTreeTest(
      '<div style="font-size: 18px"></div>'
    );
    expect(ttree).toMatchObject({
      type: 'block',
      isAnchor: false,
      attributes: {},
      styles: {
        nativeTextFlow: { fontSize: 18 }
      },
      children: []
    });
    expect(ttree?.attributes).toEqual({});
  });
});
