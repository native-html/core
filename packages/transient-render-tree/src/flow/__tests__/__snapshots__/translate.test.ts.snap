// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateNode function should comply with RFC002 example (translating) 1`] = `
TPhrasingAnchor {
  "attributes": Object {
    "href": "https://domain.com",
  },
  "children": Array [
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": "
This is
",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "text",
    },
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": "phrasing content",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "span",
      "type": "text",
    },
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": "
",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "text",
    },
    TBlock {
      "attributes": Object {
        "src": "https://domain.com/logo.jpg",
      },
      "children": Array [],
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "img",
      "type": "block",
    },
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": "
    and this is ",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "text",
    },
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": "too",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "strong",
      "type": "text",
    },
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": ".
",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "text",
    },
  ],
  "hasWhiteSpaceCollapsingEnabled": true,
  "href": "https://domain.com",
  "isAnchor": true,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {},
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "a",
  "type": "phrasing",
}
`;
